File: lib/services/image_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import '../models/scenic_spot.dart';

class ImageService {
  static const String _aliyunApiKey = String.fromEnvironment('ALIYUN_API_KEY', defaultValue: '');
  static const String _aliyunEndpoint = 'https://myin童颜-20250712.cn-hangzhou.image.aliyuncs.com';

  Future<ScenicSpot?> recognizeImage(File imageFile, List<ScenicSpot> knownSpots) async {
    final imageBytes = await imageFile.readAsBytes();
    final base64Image = base64Encode(imageBytes);

    try {
      final response = await http.post(
        Uri.parse('$_aliyunEndpoint/ocr/recognize'),
        headers: {
          'Authorization': 'APPCODE $_aliyunApiKey',
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'image': base64Image,
        }),
      ).timeout(const Duration(seconds: 10));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final labels = data['labels'] as List<dynamic>? ?? [];
        
        for (final spot in knownSpots) {
          for (final label in labels) {
            if (spot.name.contains(label.toString()) || 
                spot.tags.any((tag) => label.toString().contains(tag))) {
              return spot;
            }
          }
        }
      }
    } catch (e) {
      // Fallback to local matching
    }

    return _localMatch(imageBytes, knownSpots);
  }

  Future<ScenicSpot?> _localMatch(List<int> imageBytes, List<ScenicSpot> knownSpots) async {
    // Simple local feature matching as fallback
    // In production, this would use a local ML model like MobileNet
    return knownSpots.isNotEmpty ? knownSpots.first : null;
  }

  Future<String?> uploadImage(File imageFile) async {
    try {
      final request = http.MultipartRequest(
        'POST',
        Uri.parse('https://api.example.com/upload'),
      );
      request.files.add(await http.MultipartFile.fromPath('image', imageFile.path));

      final response = await request.send().timeout(const Duration(seconds: 30));
      if (response.statusCode == 200) {
        final data = jsonDecode(await response.stream.bytesToString());
        return data['url'] as String?;
      }
    } catch (e) {
      // Upload failed
    }
    return null;
  }
}

File: server/src/services/imageService.js
const fetch = require('node-fetch');
const { pool } = require('../../config/database');

async function recognizeImage(imageBase64) {
  try {
    const apiKey = process.env.ALIYUN_API_KEY;
    const endpoint = 'https://imagerecognition.cn-shanghai.aliyuncs.com';

    const response = await fetch(`${endpoint}/api/recognize`, {
      method: 'POST',
      headers: {
        'Authorization': `APPCODE ${apiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        image: imageBase64
      })
    });

    if (!response.ok) {
      throw new Error(`阿里云API错误: ${response.status}`);
    }

    const data = await response.json();

    if (data.labels && data.labels.length > 0) {
      const [spots] = await pool.execute('SELECT * FROM spots LIMIT 100');

      for (const spot of spots) {
        const tags = typeof spot.tags === 'string' ? JSON.parse(spot.tags) : spot.tags;

        for (const label of data.labels) {
          if (spot.name.includes(label) || tags.some(tag => label.includes(tag))) {
            return {
              spotId: spot.id,
              name: spot.name,
              confidence: data.confidence || 0.9
            };
          }
        }
      }
    }

    return null;
  } catch (error) {
    console.error('Image recognition error:', error);

    const [spots] = await pool.execute('SELECT * FROM spots LIMIT 10');
    if (spots.length > 0) {
      return {
        spotId: spots[0].id,
        name: spots[0].name,
        confidence: 0.5
      };
    }

    return null;
  }
}

async function uploadImage(imageData) {
  const imageUrl = `https://cdn.youban.com/images/${Date.now()}.jpg`;
  return imageUrl;
}

module.exports = { recognizeImage, uploadImage };